core/src/test/java/com/linecorp/armeria/client/proxy/ProxyClientIntegrationTest.java

```diff
@@ -0,0 +1,523 @@
+                                             .build();
+        final CompletableFuture<AggregatedHttpResponse> responseFuture =
+                webClient.get(PROXY_PATH).aggregate();
+        final AggregatedHttpResponse response = responseFuture.join();
+        assertThat(response.status()).isEqualByComparingTo(OK);
+        assertThat(response.contentUtf8()).isEqualTo(SUCCESS_RESPONSE);
+    }
+
+    @Test
+    void testHttpsProxyBasicCase() throws Exception {
+        final ClientFactory clientFactory =
+                ClientFactory.builder().tlsNoVerify().proxyConfig(
+                        ProxyConfig.connect(httpsProxyServer.address(), true)).build();
+        final WebClient webClient = WebClient.builder(SessionProtocol.H1C, backendServer.httpEndpoint())
+                                             .factory(clientFactory)
+                                             .decorator(LoggingClient.newDecorator())
+                                             .build();
+        final CompletableFuture<AggregatedHttpResponse> responseFuture =
+                webClient.get(PROXY_PATH).aggregate();
+        final AggregatedHttpResponse response = responseFuture.join();
+        assertThat(response.status()).isEqualByComparingTo(OK);
+        assertThat(response.contentUtf8()).isEqualTo(SUCCESS_RESPONSE);
+    }
+
+    @Test
+    void testProxyWithH2C() throws Exception {
+        final int numRequests = 5;
+        final ClientFactory clientFactory = ClientFactory.builder().proxyConfig(
+                ProxyConfig.socks4(socksProxyServer.address())).build();
+        final WebClient webClient = WebClient.builder(SessionProtocol.H2C, backendServer.httpEndpoint())
+                                             .factory(clientFactory)
+                                             .decorator(LoggingClient.newDecorator())
+                                             .build();
+
+        final List<CompletableFuture<AggregatedHttpResponse>> responseFutures = new ArrayList<>();
+        for (int i = 0; i < numRequests; i++) {
+            responseFutures.add(webClient.get(PROXY_PATH).aggregate());
+        }
+        await().until(() -> responseFutures.stream().allMatch(CompletableFuture::isDone));
+        assertThat(responseFutures.stream().map(CompletableFuture::join))
+                .allMatch(response -> response.contentUtf8().equals(SUCCESS_RESPONSE));
+    }
+
+    @Test
+    void testProxyWithUserName() throws Exception {
+        final String username = "username";
+        SOCKS_DYNAMIC_HANDLER.setChannelReadCustomizer((ctx, msg) -> {
+            if (msg instanceof DefaultSocks4CommandRequest) {
+                assertThat(username.equals(((DefaultSocks4CommandRequest) msg).userId()));
+            }
+            ctx.fireChannelRead(msg);
+        });
+
+        final ClientFactory clientFactory = ClientFactory.builder().proxyConfig(
+                ProxyConfig.socks4(socksProxyServer.address(), username)).build();
```

Comment:
nit: Indent for readability
```suggestion
        final ClientFactory clientFactory =
                ClientFactory.builder()
                             .proxyConfig(ProxyConfig.socks4(socksProxyServer.address(), username))
                             .build();
```